### ISSUE TITLE
Queries with more than 32k parameters are allowed

### ISSUE THREAD

**Issue by flaker** [brianc/node-postgres#2579]
version: 8.0.3

As far as I understand, a query can have up to max int16 parameters (from https://www.postgresql.org/docs/devel/protocol-message-formats.html). I found out in my logs error messages like the following.

```
ERROR:  bind message has 52708 parameter formats but 0 parameters
```

The means that the prepare message was fine expecting 52708 parameters? and if that's the case how would that be handled? that takes me the "0 parameters". I am plainly executing a query, how is it that 0 parameters showed up?

Thank you.

**Comment by vitaly-t** [comment]
That error message means you've got 52708 placeholders, but no values passed in.

The maximum is 65535, which the only way you can hit if you are using it the wrong way. You should be splitting those across queries or hold inside an array, not bundle all into one query.

**Comment by brianc** [comment]
LMK if you have code to reproduce this.  Max int16 value is max number of params supported by the postgres protocol. I think @vitaly-t is right here you aren't passing params properly.

**Comment by flaker** [comment]
@brianc it's not code I wrote. I also think it's the incorrect way to do it. Is it uint16 or signed int16?

Your comment here:
https://github.com/brianc/node-postgres/issues/2292#issuecomment-663639630

indicate a signed int16

**Comment by brianc** [comment]
yeah sorry you're right it's signed.  I'm not following what the issue is related to _exactly_ - do you have code to reproduce?

**Comment by flaker** [comment]
Hi @brianc ,

Sorry, probably poor wording on my side. The issue is the following. This application has been running for a while sending (some times) more than 32k positional arguments. AFAIK these queries are being prepared (as in strings and parameters not actual posgresql "prepare message") by a library called Squel.js. And later own plainly executed via this module. Likely these calls have been failing and silently been put on a log somewhere.

My concerns (the issue) are:

 - If the limit to arguments is 32k, how is it possible that we get to the bind message to fail at that stage? Is it something that Posgresql does? allowing to prepare a statement that has no chance of actually executing? I am assuming the extended query flow here. Maybe some error check in node-postgres is missing? or maybe a preemptive check on the size of parameters?

 - If postgresql allows (and require) to "prepare" a broken statement (> 32k parameters). Node-postgres will attempt to send them all? I think I've seen some byte shuffling around the parameters array size. So if I send 40k parameters, actually the library will send a bind message with a completely different number (https://github.com/brianc/node-postgres/blob/4b229275cfe41ca17b7d69bd39f91ada0068a5d0/packages/pg-protocol/src/serializer.ts#L155). So, allowing more than 32k parameters might end up in a completely unintended behavior.

Again, I also think it's a bad idea to send so many parameters, but found it odd that the library would allow it and postgresql will inform at such a later stage.

Thanks.
